src/inc/til/u8u16convert.h

```diff
@@ -386,14 +247,69 @@ namespace til // Terminal Implementation Library. Also: "Today I Learned"
     // - E_OUTOFMEMORY - the function failed to allocate memory for the resulting string
     // - E_ABORT       - the resulting string length would exceed the upper boundary of an int and thus, the conversion was aborted before the conversion has been completed
     // - E_UNEXPECTED  - an unexpected error occurred
-    template<class inT, class outT>
-    [[nodiscard]] typename std::enable_if<std::is_same<typename inT::value_type, wchar_t>::value && std::is_same<typename outT::value_type, char>::value, HRESULT>::type
-    u16u8(const inT in, outT& out, u16state& state) noexcept
+    template<class outT>
+    [[nodiscard]] HRESULT u16u8(const std::wstring_view& in, outT& out, u16state& state) noexcept
     {
-        std::wstring_view sv{};
-        RETURN_IF_FAILED(state(std::wstring_view{ in }, sv));
-        return u16u8(sv, out);
+        try
+        {
+            out.clear();
+            RETURN_HR_IF(S_OK, in.empty());
+
+            int len16{};
+            int capa8{};
+            // The worst ratio of UTF-16 code units to UTF-8 code units is 1 to 3.
+            RETURN_HR_IF(E_ABORT, !base::MakeCheckedNum(in.length()).AssignIfValid(&len16) || !base::CheckAdd(len16, gsl::narrow_cast<int>(state.partials[0]) != 0).AssignIfValid(&capa8) || !base::CheckMul(capa8, 3).AssignIfValid(&capa8));
+
+            out.resize(gsl::narrow_cast<size_t>(capa8));
+            int len8{};
+            auto cursor16{ in.data() };
+            if (state.partials[0])
+            {
+                state.partials[1] = *cursor16;
+                len8 = WideCharToMultiByte(CP_UTF8, 0UL, &state.partials[0], 2, out.data(), capa8, nullptr, nullptr);
+                RETURN_HR_IF(E_UNEXPECTED, !len8);
+
+                state.reset();
+                capa8 -= len8;
+                --len16;
+                ++cursor16;
+            }
+
+            if (len16)
+            {
+                const auto back = *(cursor16 + len16 - 1);
+                if (back >= 0xD800 && back <= 0xDBFF) // cache the last value in the string if it is in the range of high surrogates
+                {
+                    state.partials[0] = back;
+                    --len16;
+                }
+            }
+
+            if (len16)
+            {
+                const auto convLen{ WideCharToMultiByte(CP_UTF8, 0UL, cursor16, len16, out.data() + len8, capa8, nullptr, nullptr) };
+                RETURN_HR_IF(E_UNEXPECTED, !convLen);
+
+                len8 += convLen;
+            }
+
+            out.resize(gsl::narrow_cast<size_t>(len8));
+            return S_OK;
+        }
+        catch (std::length_error&)
+        {
+            return E_ABORT;
+        }
+        catch (std::bad_alloc&)
+        {
+            return E_OUTOFMEMORY;
+        }
+        catch (...)
+        {
+            return E_UNEXPECTED;
+        }

```

Comment:
Just a nit: The two instances of this code can be replaced with the [`CATCH_RETURN();` macro](https://github.com/Microsoft/wil/wiki/Error-handling-helpers#exception-guards), which handles the same cases.